scheds/rust/scx_flash/src/bpf/main.bpf.c

```diff
@@ -110,322 +342,1422 @@ int kprobe_vfs_fsync_range(struct file *file, u64 start, u64 end, int datasync)
+		 * shares the same L2 cache.
+		 */
+		if (l2_mask) {
+			cpu = __COMPAT_scx_bpf_pick_idle_cpu_node(l2_mask, node,
+						SCX_PICK_IDLE_CORE | __COMPAT_SCX_PICK_IDLE_IN_NODE);
+			if (cpu >= 0) {
+				*is_idle = true;
+				goto out_put_cpumask;
+			}
+		}
+
+		/*
+		 * Search for any full-idle CPU in the primary domain that
+		 * shares the same L3 cache.
+		 */
+		if (l3_mask) {
+			cpu = __COMPAT_scx_bpf_pick_idle_cpu_node(l3_mask, node,
+						SCX_PICK_IDLE_CORE | __COMPAT_SCX_PICK_IDLE_IN_NODE);
+			if (cpu >= 0) {
+				*is_idle = true;
+				goto out_put_cpumask;
+			}
+		}
+
+		/*
+		 * Search for any full-idle CPU in the primary domain.
+		 *
+		 * If the current node needs a rebalance, look for any
+		 * full-idle CPU also on different nodes.
+		 */
+		if (p_mask) {
+			u64 flags = SCX_PICK_IDLE_CORE;
+
+			if (!node_rebalance(node))
+				flags |= __COMPAT_SCX_PICK_IDLE_IN_NODE;
+
+			cpu = __COMPAT_scx_bpf_pick_idle_cpu_node(p_mask, node, flags);
+			if (cpu >= 0) {
+				*is_idle = true;
+				goto out_put_cpumask;
+			}
+		}
+	}
+
+	/*
+	 * If a full-idle core can't be found (or if this is not an SMT system)
+	 * try to re-use the same CPU, even if it's not in a full-idle core.
+	 */
+	if (is_prev_allowed &&
+	    scx_bpf_test_and_clear_cpu_idle(prev_cpu)) {
+		cpu = prev_cpu;
+		*is_idle = true;
+		goto out_put_cpumask;
+	}
+
+	/*
+	 * Search for any idle CPU in the primary domain that shares the same
+	 * L2 cache.
+	 */
+	if (l2_mask && !node_rebalance(node)) {
+		cpu = __COMPAT_scx_bpf_pick_idle_cpu_node(l2_mask, node,
+						__COMPAT_SCX_PICK_IDLE_IN_NODE);
+		if (cpu >= 0) {
+			*is_idle = true;
+			goto out_put_cpumask;
+		}
+	}
+
+	/*
+	 * Search for any idle CPU in the primary domain that shares the same
+	 * L3 cache.
+	 */
+	if (l3_mask && !node_rebalance(node)) {
+		cpu = __COMPAT_scx_bpf_pick_idle_cpu_node(l3_mask, node,
+						__COMPAT_SCX_PICK_IDLE_IN_NODE);
+		if (cpu >= 0) {
+			*is_idle = true;
+			goto out_put_cpumask;
+		}
+	}
+
+	/*
+	 * Search for any idle CPU in the scheduling domain.
+	 */
+	if (p_mask) {
+		cpu = __COMPAT_scx_bpf_pick_idle_cpu_node(p_mask, node, 0);
+		if (cpu >= 0) {
+			*is_idle = true;
+			goto out_put_cpumask;
+		}
+	}
+
+	/*
+	 * Search for any idle CPU usable by the task.
+	 */
+	cpu = __COMPAT_scx_bpf_pick_idle_cpu_node(p->cpus_ptr, node, 0);
+	if (cpu >= 0) {
+		*is_idle = true;
+		goto out_put_cpumask;
+	}
+
+out_put_cpumask:
+	scx_bpf_put_cpumask(idle_cpumask);
+	scx_bpf_put_cpumask(idle_smtmask);
+
+	/*
+	 * If we couldn't find any CPU, or in case of error, return the
+	 * previously used CPU.
+	 */
+	if (cpu < 0)
+		cpu = prev_cpu;
+
+	return cpu;
+}
+
+/*
+ * Return true if we can perform a direct dispatch on @node, false
+ * otherwise.
+ */
+static bool can_direct_dispatch(int node)
+{
+	/*
+	 * Allow direct dispatch when @local_pcpu is enabled, or when there
+	 * are no tasks queued in the node DSQ.
+	 */
+	return local_pcpu || !scx_bpf_dsq_nr_queued(node);
```

Comment:
I think it would be interesting to test a little less conservative check on nr_queued (especially on big machines). For smaller single node machines this probably makes sense though.